### DISCUSSION TITLE
Why is this code working? Didn't see it in docs

### DISCUSSION THREAD

**Discussion by cooperinho** [yajra/laravel-datatables#2749] [Q&A] [answered]
I am working on an existing code and I didn't find any explanation about the way this code is written in the docs, so I wonder how it works and how the previous developer came up with this :

```
<?php

namespace App\DataTables;

class MyDataTable extends DataTable
{
    /**
     * Build DataTable class.
     *
     * @param mixed $query Results from query() method.
     * @return \Yajra\DataTables\DataTableAbstract
     */
    public function dataTable($query)
    {
        return datatables()
            ->($query);       
    }

    /**
     * Get query source of dataTable.
     *
     * @param \App\Models\MyModel $model
     * @return \Illuminate\Database\Eloquent\Builder
     */
    public function query(MyModel $model)
    {
        return $model:select('*')...
    }

    /**
     * Optional method if you want to use html builder.
     *
     * @return \Yajra\DataTables\Html\Builder
     */
    public function html()
    {
        return $this->builder()
            ->setTableId('my-datatable-id')
            ->columns($this->getColumns())
            ->dom('Bfrtip')
            ->orderBy(1)
	    ->ajax([
                'type' => 'GET',
                'url' => route('controller_method_that_renders_datatable')
            ]);;

    }

    /**
     * Get columns.
     *
     * @return array
     */
    protected function getColumns()
    {
        return [
            ['title' => 'ID', 'data' => 'id'],
            ['title' => 'Status', 'data' => 'status'],
            ['title' => 'Account Number', 'data' => 'number'],
            ['title' => 'Created By', 'data' => 'created_by'],
            ['title' => 'Created At', 'data' => "created_at"],

        ];
    }

}

```

This is how the table is being called in the controller:
```
    public function controller_method_that_renders_datatable(MyDataTable $myDataTable)
    {
        return $myDataTable->render('path.to.datatable.view');
    }
```

and in the blade view it's just:

```
{!! $myDataTable->table() !!}
```

Oh and in the `index` method of the controller the general blade view is being called:

```
    public function index(MyDataTable $myDataTable) 
    {
        return view('path.to.datatable.view',['myDataTable' => $myDataTable->html()]);
    }
```

Things I didn't understand for example:

- Why is the `->ajax()` method used? without it it doesn't work. At first I tried to follow the docs but nothing worked, until I compared my code to an existing datatable code in the project and realized that it's working when adding the `ajax` method
- Why I can set the columns title using this array syntax (`['title' => 'ID', 'data' => 'id']`) - I didn't find it in the docs, but it works!
- Why the datatable is being called twice in the controller? once with `->html()` from the `index` method and then from the specific method `controller_method_that_renders_datatable`??

This data table is in my Laravel's project DataTables folder. Does this implementation look correct? Or there are some stuff that I should change?

**Comment by yajra** [comment, marked as answer]
> Why is the ->ajax() method used? without it it doesn't work. At first I tried to follow the docs but nothing worked, until I compared my code to an existing datatable code in the project and realized that it's working when adding the ajax method

You should use the dataTables render function to be able to handle multiple actions/request:

```php
$dataTable->render('view.path')
```

1. It will render the view on regular http GET request
2. It will render json data if requested via ajax
3. It will handle the export (xls, csv, pdf), print and custom table actions

> Why I can set the columns title using this array syntax (['title' => 'ID', 'data' => 'id']) - I didn't find it in the docs, but it works!

You can use an [array](https://yajrabox.com/docs/laravel-datatables/master/html-builder-column) or the fluent Column class builder. The fluent builder is not fully documented but it's similar to array version but instead of keys, use the counter part method.

> Why the datatable is being called twice in the controller? once with ->html() from the index method and then from the specific method controller_method_that_renders_datatable??

You should only use render method.
